src/workflow/snakemake/format/formatter.py

```
@@ -12,16 +17,43 @@ def format_output_templates_to_be_expanded(node):
     return outputs
 
 
-def extract_stages_from_path(path, known_stages):
+def format_input_templates_to_be_expanded(benchmark: Benchmark, wildcards):
+    """Formats benchmark inputs that will be expanded according to Snakemake's engine"""
+
+    pre = wildcards.pre
+    post = wildcards.post
+    name = wildcards.name
+
+    nodes = benchmark.get_nodes()
+    stage_ids = set([node.stage_id for node in nodes])
+
+    pre_stages = _extract_stages_from_path(pre, stage_ids)
+    after_stage = pre_stages[-1][0] if len(pre_stages) > 0 else None
+
+    stage_id, module_id, param_id = _match_node_format(post)
+
+    node_hash = hash(BenchmarkNode.to_id(stage_id, module_id, param_id, after_stage))
+    matching_node = next((node for node in nodes if hash(node) == node_hash), None)
+    assert matching_node is not None
```

__Fix__: It is better to raise an error with a clear error message, than using `assert`. I think that is mainly used for tests

---

Agree, took some shortcuts during initial development, and originally the asserts were just to make sure i'm covering all the corner cases. At this point they can either be entirely removed or transformed into an error. I will make them a runtime error.